File: BOJ/1517.cpp
#include <iostream>
#include <vector>

#define MAX 500000
using namespace std;

vector<int> v;
int N;
long long ans=0;

void merge(int start, int mid, int end, int left, int right);
void mergeSort(int start,int end);
int main(){
    // 첫째 줄에 N(1 ≤ N ≤ 500,000)이 주어진다.
    cin >> N;
    // 다음 줄에는 N개의 정수로 A[1], A[2], …, A[N]이 주어진다.
    // 각각의 A[i]는 0 ≤ |A[i]| ≤ 1,000,000,000의 범위에 들어있다.
    // int 최대값-2,147,483,648 ~ 2,147,483,647 범위 내
    for (int i=0; i<N; i++){
        int tmp;
        cin >> tmp;
        v.push_back(tmp);
    }
    mergeSort(0, N-1);
    cout << ans << '\n';
    return 0;
}
void mergeSort(int start,int end){
    int mid = (start+end)/2;
    int left=start;
    int right=mid+1;

    if (start==end) return;

    mergeSort(start, mid);
    mergeSort(mid+1, end);

    merge(start, mid, end, left, right);
}

void merge(int start, int mid, int end, int left, int right){
    vector<int> tmp;
    // tmp에 크기 순서로 정렬하기
    while (left<=mid && right<=end){
        if (v[left]<=v[right]){
            tmp.push_back(v[left++]);
            // 아래 코드와 동일한 코드
            // tmp.push_back(v[left]);
            // left++;
        }
        // 앞뒤 순서가 바뀌어야 하는 상태, 두 수가 떨어진 거리만큼 ans에 더하기
        else{
            tmp.push_back(v[right++]);
            ans+=(mid+1-left);
        }
    }
    while (left<=mid){
        tmp.push_back(v[left++]);
    }
    while (right<=end){
        tmp.push_back(v[right++]);
    }

    // tmp에 정렬한 순서로 원본 벡터 v 변경해주기
    for (int i=start,j=0; i<=end; i++,j++){
        v[i]=tmp[j];
    }
}

File: BOJ JAVA/1517.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    public static long cnt = 0;
    public static long[] sorted;

    public static void main(String[] args) throws IOException {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in))) {
            int n = Integer.parseInt(br.readLine());
            sorted = new long[n];
            long[] arr = new long[n];

            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < n; i++) {
                arr[i] = Long.parseLong(st.nextToken());
            }

            // n개의 원소 mergesort
            mergeSort(arr, 0, n - 1);
            System.out.println(cnt);
        }
    }

    // 쪼갠 다음 합치면서 정렬
    public static void mergeSort(long[] arr, int start, int end) {
        if (start == end) return;

        int mid = (start + end) / 2;

        mergeSort(arr, start, mid);
        mergeSort(arr, mid + 1, end);
        merge(arr, start, mid, end);
    }

    // 실제로 합쳐주는 작업
    public static void merge(long[] arr, int start, int mid, int end) {
        int left = start;
        int right = mid + 1;
        int idx = start;

        while (left <= mid && right <= end) {
            if (arr[left] <= arr[right]) {
                sorted[idx++] = arr[left++];
            } else {
                sorted[idx++] = arr[right++];
                cnt += (mid + 1 - left);
            }
        }

        while (left <= mid) {
            sorted[idx++] = arr[left++];
        }

        while (right <= end) {
            sorted[idx++] = arr[right++];
        }

        // 정렬된 결과를 원본 배열에 반영
        for (int i = start; i <= end; i++) {
            arr[i] = sorted[i];
        }
    }
}
